src/main/scala/is/hail/methods/MaximalIndependentSet.scala

```diff
@@ -40,42 +35,45 @@ object MaximalIndependentSet {
       Graph(toBeComputed.vertices.leftZipJoin(toBeComputed.degrees) { (v, _, degree) => (degree.getOrElse(0), v) }, toBeComputed.edges)
     }
 
-    var graph = updateVertexDegrees(g)
-    val edgeDirection = if (undirected) EdgeDirection.Either else EdgeDirection.Out
+    val originalGraph: Graph[Message, ED] = updateVertexDegrees(g)
+    originalGraph.persist()
 
-    while(graph.numEdges > 0) {
-      graph = graph.pregel(initialMsg, Int.MaxValue, edgeDirection)(receiveMessage, sendMsg, mergeMsg)
-      graph = graph.subgraph(_ => true, (id, value) => value match { case (maxDegrees, maxID) => maxID != id || maxDegrees == 0})
-      graph = updateVertexDegrees(graph)
-    }
-    graph.vertices.keys.collect().toSet
-  }
+    val edgeDirection = if (undirected) EdgeDirection.Either else EdgeDirection.Out
 
+    var idSetToDelete = Set[VertexId]()
+    var hasEdges = true
 
-  def ofIBDMatrix(sc: SparkContext, inputRDD: RDD[((String, String), Double)], thresh: Double): Set[String] = {
-    //Filter RDD to remove edges above threshold
-    val filteredRDD = inputRDD.filter(_._2 <= thresh)
+    while (hasEdges) {
+      val filteredGraph = updateVertexDegrees(originalGraph.subgraph(_ => true, (id, value) => !idSetToDelete.contains(id)))
+      filteredGraph.persist()
+      
+      if(filteredGraph.numEdges == 0) {
+        hasEdges = false
+      }
 
-    //Throw away weights
-    val vertexPairs = inputRDD.keys
+      val pregelGraph = filteredGraph.pregel(initialMsg, Int.MaxValue, edgeDirection)(receiveMessage, sendMsg, mergeMsg)
 
-    //Collect all vertices.
-    val allVertices = vertexPairs.flatMap[String]{case (v1, v2) => List(v1, v2)}.distinct()
+      idSetToDelete = idSetToDelete.union(pregelGraph.vertices
+        .filter(tuple => tuple match {case (id, value) => value match  {case (maxDegrees, maxID) => maxID == id && maxDegrees != 0}})
+        .map(_._1).collect().toSet)
 
-    val numberedVertices: RDD[(String, VertexId)] = allVertices.zipWithIndex()
+      filteredGraph.unpersist()
+    }
+    originalGraph.subgraph(_ => true, (id, value) => !idSetToDelete.contains(id)).vertices.keys.collect().toSet
+  }
 
-    val verticesToNumbers = numberedVertices.collectAsMap()
-    val numbersToVertices = numberedVertices.map{case (name, id) => (id, name)}.collectAsMap()
 
+  def ofIBDMatrix(inputRDD: RDD[((Int, Int), Double)], thresh: Double, vertexIDs: Seq[Int]): Set[Long] = {
```

Comment:
We cannot. In the line 

`val edges: RDD[Edge[Double]] = filteredRDD.map{case((v1, v2), weight) => Edge(v1, v2, weight)}`

`v1` and `v2` are the integers being used as ids for the vertices. Technically longs would also work, but only ints or longs.